src/main/scala/io/iohk/ethereum/ledger/BlockImport.scala

```diff
@@ -0,0 +1,260 @@
+package io.iohk.ethereum.ledger
+
+import akka.util.ByteString
+import io.iohk.ethereum.consensus.validators.BlockHeaderError.HeaderParentNotFoundError
+import io.iohk.ethereum.domain._
+import io.iohk.ethereum.ledger.BlockExecutionError.ValidationBeforeExecError
+import io.iohk.ethereum.ledger.BlockQueue.Leaf
+import io.iohk.ethereum.metrics.{ Metrics, MetricsClient }
+import io.iohk.ethereum.utils.{ BlockchainConfig, Logger }
+import org.bouncycastle.util.encoders.Hex
+
+import scala.annotation.tailrec
+import scala.concurrent.{ ExecutionContext, Future }
+
+class BlockImport(
+  blockchain: BlockchainImpl,
+  blockQueue: BlockQueue,
+  blockchainConfig: BlockchainConfig,
+  executeBlock: (Block, Boolean) => Either[BlockExecutionError, Seq[Receipt]],
+  validateBlockBeforeExecution: Block => Either[ValidationBeforeExecError, BlockExecutionSuccess]
+)(implicit blockExecutionContext: ExecutionContext) extends Logger {
+
+  private[ledger] def importToTop(
+    block: Block,
+    currentBestBlock: Block,
+    currentTd: BigInt
+  )(implicit blockExecutionContext: ExecutionContext): Future[BlockImportResult] = {
+    val validationResult = Future(validateBlockBeforeExecution(block))
+
+    val importResult = Future(importBlockToTop(block, currentBestBlock.header.number, currentTd))
+
+    for {
+      valRes    <- validationResult
+      importRes <- importResult
+    } yield {
+      valRes.fold(error => handleImportTopValidationError(error, block, currentBestBlock, importRes), _ => importRes)
+    }
+  }
+
+  private def importBlockToTop(
+    block: Block,
+    bestBlockNumber: BigInt,
+    currentTd: BigInt
+  )(implicit blockExecutionContext: ExecutionContext): BlockImportResult = {
+    val topBlockHash = blockQueue.enqueueBlock(block, bestBlockNumber).get.hash
+    val topBlocks = blockQueue.getBranch(topBlockHash, dequeue = true)
+    val (importedBlocks, maybeError) = executeBlocks(topBlocks, currentTd)
+
+    val result = maybeError match {
+      case None =>
+        BlockImportedToTop(importedBlocks)
+
+      case Some(error) if importedBlocks.isEmpty =>
+        blockQueue.removeSubtree(block.header.hash)
+        BlockImportFailed(error.toString)
+
+      case Some(_) =>
+        topBlocks.drop(importedBlocks.length).headOption.foreach { failedBlock =>
+          blockQueue.removeSubtree(failedBlock.header.hash)
+        }
+        BlockImportedToTop(importedBlocks)
+    }
+
+    importedBlocks.foreach { b =>
```

Comment:
This loop should be inside `debug` and lazy evaluated.

Reply:
you mean something like this: 
```
log.debug{
        lazy val result = importedBlocks.map { b =>
          val header = b.block.header
        s"Imported new block (${header.number}: ${Hex.toHexString(header.hash.toArray)}) to the top of chain \n"
        }
        result.toString
      }
```

Reply:
It still won't be lazy (I think).
```
log.debug("{}", {
 lazy val result = importedBlocks.map { b =>
 val header = b.block.header
 s"Imported new block (${header.number}: ${Hex.toHexString(header.hash.toArray)}) to the 
 top of chain \n"
 }
 result.toString
})
```